runtime/runtime/src/congestion_control.rs

```diff
@@ -256,12 +275,19 @@ impl ReceiptSinkV2<'_> {
         // could be a special case during resharding events. Or even a bug. In
         // any case, if we cannot know a limit, treating it as literally "no
         // limit" is the safest approach to ensure availability.
-        let forward_limit = outgoing_limit.entry(shard).or_insert(Gas::MAX);
+        // For the size limit, we default to the usual limit that is applied to all (non-special) shards.
+        let forward_limit = outgoing_limit.entry(shard).or_insert(OutgoingLimit {
+            gas: Gas::MAX,
+            size: apply_state.config.witness_config.outgoing_receipts_usual_size_limit,
+        });
         let gas_to_forward = receipt_congestion_gas(&receipt, &apply_state.config)?;
-        if *forward_limit > gas_to_forward {
+        let size_to_forward = receipt_size(&receipt)?;
+
+        if forward_limit.gas > gas_to_forward && forward_limit.size > size_to_forward {
```

Comment:
sanity check - We should either make really sure that the special forward limit is larger than maximum potential receipt size or make it a soft limit in the special case. I think the former is already the case but just making sure in case the receipt size is a soft limit or some other subtlety.

Reply:
`outgoing_receipts_big_size_limit` is 4.5MiB, which was chosen to be a bit above 4MiB to allow 4MiB receipts to pass. It's a hard limit, so it has to be a bit bigger.

But now I'm starting to have doubts, do we have a size limit for receipts? I thought that there'd be something analogous to `max_transaction_size`, but I don't see it 0_o. Maybe we'll have to add `max_receipt_size` limit as well.

Reply:
Hehe I was aiming for doubts :) 
The issue is that just one receipt over the limit would block the queue forever. Making it a soft limit for the special shard should solve this problem but there may be better alternatives. Are large receipts a more general problem that we need to solve or is it only an issue in cross shard receipts?

Reply:
I am not entirely sure but I believe a single receipts can be much larger than 4MB today

To quote past Jakob, writing about potentially introducing a receipt size limit:

> ~5MB for bandwidth per receipts seems reasonable to me. A receipt can produce lots of network bandwidth by making a cross contract call. Today, it's possible to make many calls at once and attach an argument of up to 4MB to each. We could say the total limit per receipt is 5MB, then users can still make a single cross contract call with a 4MB argument, or two calls with 2MB each. But they couldn't do two calls with 4MB each. I don't think that's a problem for anyone.

source: https://github.com/near/nearcore/issues/9378#issuecomment-1700964833

Based on that, a receipt could be around 100MB if it contains 25 function calls with 4 MB payload each. The gas cost for creating all those actions would only be below 250 Tgas, so plenty of room for the initial call and promise overhead costs.

Reply:
> Based on that, a receipt could be around 100MB if it contains 25 function calls with 4 MB payload each. The gas cost for creating all those actions would only be below 250 Tgas, so plenty of room for the initial call and promise overhead costs.

Damn, that's not good. I guess we'll have to add a receipt size limit then. I hope it's safe to do it, rejecting receipts sounds dangerous 0_O

Reply:
Oh my I sparked more doubts than I was hoping for. If the receipts can be so large then indeed even a soft limit would not be sufficient. @jancionear Do you want to share the news on zulip or at the sync meeting later?

Reply:
https://near.zulipchat.com/#narrow/stream/295306-contract-runtime/topic/How.20large.20can.20internal.20receipts.20get.3F/near/443077081